### DISCUSSION TITLE
Support for YARL and Hyperlink URL objects

### DISCUSSION THREAD

**Discussion by gwerbin** [encode/httpx#1912] [Ideas]
It'd be really nice if HTTPX suported [`yarl.URL`](https://yarl.readthedocs.io/) and [`hyperlink.EncodedURL`/`hyperlink.DecodedURL`](https://hyperlink.readthedocs.io/) in addition to `httpx.URL`. The data models are pretty similar, and I think it should be possible to convert directly from any of these types to `httpx.URL` without round-tripping through `str` first.

Would the maintainers be open to a PR for this?

If so, would `httpx.Request.__init__` be the right place to do this conversion? That seems to be the "entry point" where all non-`URL` objects are converted to `URL`: <https://github.com/encode/httpx/blob/1.0.0.beta0/httpx/_models.py#L1092>. For example, we could have a `_convert_url` helper function instead of calling `URL` directly.

Naturally, it would need to gracefully handle the case when either package is not available, and the type annotations would have to be updated.

**Comment by gwerbin** [comment]
I've been using HTTPX a lot and I'd still be very interested in helping contribute here.

**Reply by karpetrosyan** [reply]
Is it truly necessary? Adding such features is not cheap; it complicates the development process, introduces new dependencies, and so on.

**Reply by gwerbin** [reply]
No, it's definitely not _necessary_.
